pluginframework/common/tf_to_go.go

```diff
@@ -0,0 +1,220 @@
+		return nil
+	}
+
+	if !destField.CanSet() {
+		panic(fmt.Errorf("destination field can not be set: %s", destField.Type().Name()))
+	}
+	srcFieldValue := srcField.Interface()
+
+	if srcFieldValue == nil {
+		return nil
+	} else if srcField.Kind() == reflect.Ptr {
+		if srcField.IsNil() {
+			// Skip nils
+			return nil
+		}
+		// Allocate new memory for the destination field
+		destField.Set(reflect.New(destField.Type().Elem()))
+
+		// Recursively populate the nested struct.
+		if TfSdkToGoSdkStruct(srcFieldValue, destField.Interface(), ctx).ErrorsCount() > 0 {
+			panic("Error converting tfsdk to gosdk struct")
+		}
+	} else if srcField.Kind() == reflect.Struct {
+		tfsdkToGoSdkStructField(srcField, destField, srcFieldName, forceSendFieldsField, ctx)
+	} else if srcField.Kind() == reflect.Slice {
+		if srcField.IsNil() {
+			// Skip nils
+			return nil
+		}
+		destSlice := reflect.MakeSlice(destField.Type(), srcField.Len(), srcField.Cap())
+		for j := 0; j < srcField.Len(); j++ {
+
+			srcElem := srcField.Index(j)
+
+			destElem := destSlice.Index(j)
+			if err := tfSdkToGoSdkSingleField(srcElem, destElem, "", nil, ctx); err != nil {
+				return err
+			}
+		}
+		destField.Set(destSlice)
+	} else if srcField.Kind() == reflect.Map {
+		if srcField.IsNil() {
+			// Skip nils
+			return nil
+		}
+		destMap := reflect.MakeMap(destField.Type())
+		for _, key := range srcField.MapKeys() {
+			srcMapValue := srcField.MapIndex(key)
+			destMapValue := reflect.New(destField.Type().Elem()).Elem()
+			destMapKey := reflect.ValueOf(key.Interface())
+			if err := tfSdkToGoSdkSingleField(srcMapValue, destMapValue, "", nil, ctx); err != nil {
+				return err
+			}
+			destMap.SetMapIndex(destMapKey, destMapValue)
+		}
+		destField.Set(destMap)
+	} else {
+		panic(fmt.Errorf("unknown type for field: %s", srcField.Type().Name()))
+	}
+	return nil
+
+}
+
+func tfsdkToGoSdkStructField(srcField reflect.Value, destField reflect.Value, srcFieldName string, forceSendFieldsField *reflect.Value, ctx context.Context) {
+	srcFieldValue := srcField.Interface()
+	switch v := srcFieldValue.(type) {
+	case types.Bool:
+		destField.SetBool(v.ValueBool())
+		if !v.IsNull() {
+			addToForceSendFields(srcFieldName, forceSendFieldsField)
+		}
+	case types.Int64:
+		destField.SetInt(v.ValueInt64())
+		if !v.IsNull() {
+			addToForceSendFields(srcFieldName, forceSendFieldsField)
+		}
+	case types.Float64:
+		destField.SetFloat(v.ValueFloat64())
+		if !v.IsNull() {
+			addToForceSendFields(srcFieldName, forceSendFieldsField)
+		}
+	case types.String:
+		if destField.Type().Name() != "string" {
+			// This is the case for enum.
+
+			// Skip unset value.
+			if srcField.IsZero() {
+				return
+			}
+
+			// Find the Set method
+			destVal := reflect.New(destField.Type())
+			setMethod := destVal.MethodByName("Set")
+			if !setMethod.IsValid() {
+				panic(fmt.Sprintf("set method not found on enum type: %s", destField.Type().Name()))
+			}
```

Comment:
We don't have the same case with enum pointers as above?

Reply:
Not needed because we always instantiate the memory for this field in the same way two lines above.